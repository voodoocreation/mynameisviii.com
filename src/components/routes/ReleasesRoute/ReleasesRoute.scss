@import "../../../scss/variables", "../../../scss/mixins";

.ReleasesRoute--listings {
  @include panel-1();

  margin-bottom: $spacing;
  padding: $spacing;

  h2 {
    @include line-before-after();

    margin-top: 0;
  }

  &--items {
    display: flex;
    flex-flow: row wrap;
    margin: (-$spacing-half);
  }
}
