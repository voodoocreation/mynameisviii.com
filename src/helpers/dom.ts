import getScrollbarWidth from "scrollbar-size";

if (typeof window !== "undefined") {
  window.scrollState = {
    count: 0,
    left: 0,
    top: 0
  };
}

export const isServer = () => typeof window === "undefined" || window.isServer;

export const isInViewport = (element?: HTMLElement | null) => {
  if (!element) {
    return false;
  }

  const bounds = element.getBoundingClientRect();
  const viewport = {
    height: window.innerHeight || document.documentElement.clientHeight,
    width: window.innerWidth || document.documentElement.clientWidth
  };

  return (
    bounds.top >= 0 &&
    bounds.right <= viewport.width &&
    bounds.bottom <= viewport.height &&
    bounds.left >= 0
  );
};

export const isAlmostInViewport = (
  element?: HTMLElement | null,
  offset = 0
) => {
  if (!element) {
    return false;
  }

  const bounds = element.getBoundingClientRect();
  const viewport = {
    height: window.innerHeight || document.documentElement.clientHeight,
    width: window.innerWidth || document.documentElement.clientWidth
  };

  return (
    viewport.height - bounds.top > -offset &&
    bounds.right > -offset &&
    bounds.bottom > -offset &&
    viewport.width - bounds.left > -offset
  );
};

export const lockScroll = () => {
  const { body } = document;
  const html = document.documentElement;

  window.scrollState.count += 1;

  if (!html.classList.contains("isLocked")) {
    const existingPadding = parseInt(
      getComputedStyle(body, null).getPropertyValue("padding-right"),
      10
    );
    const scrollbarWidth = getScrollbarWidth();
    const newPadding =
      isNaN(existingPadding) || !isFinite(existingPadding)
        ? scrollbarWidth
        : scrollbarWidth + existingPadding;
    const scrollElement =
      body.scrollTop > 0 || html.scrollTop === 0 ? body : html;

    window.scrollState.top = scrollElement.scrollTop;
    window.scrollState.left = scrollElement.scrollLeft;

    html.classList.add("isLocked");
    body.style.paddingRight = `${newPadding}px`;
  }
};

export const unlockScroll = () => {
  const { body } = document;
  const html = document.documentElement;

  if (window.scrollState.count <= 1 && html.classList.contains("isLocked")) {
    body.style.paddingRight = "";
    html.classList.remove("isLocked");
    window.scrollTo(window.scrollState.left, window.scrollState.top);
  }

  if (window.scrollState.count > 0) {
    window.scrollState.count -= 1;
  }
};
