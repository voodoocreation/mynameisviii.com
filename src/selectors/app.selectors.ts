import { TStoreState } from "../reducers/root.reducers";

export const getCurrentRoute = (state: TStoreState) => state.app.currentRoute;

export const getAppError = (state: TStoreState) => state.app.error;

export const isAppLoading = (state: TStoreState) => state.app.isLoading;

export const hasNewVersion = (state: TStoreState) => state.app.hasNewVersion;

export const isNavOpen = (state: TStoreState) => state.app.isNavOpen;

export const isOnline = (state: TStoreState) => state.app.isOnline;
