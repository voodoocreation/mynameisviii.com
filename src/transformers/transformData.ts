export const arrayToAssoc = (
  array: Array<{}>,
  key: string,
  failSilently = true
) =>
  array.reduce(
    (
      acc: { [index: string]: {} },
      curr: { [index: string]: any },
      index: number
    ) => {
      if (curr[key] !== undefined) {
        acc[curr[key]] = curr;
      } else if (!failSilently) {
        throw new Error(`Key '${key}' not found at index ${index}`);
      }
      return acc;
    },
    {}
  );

export const assocToArray: (object: any) => any = object =>
  Object.keys(object).reduce((acc: Array<{}>, curr: string) => {
    acc.push(object[curr]);
    return acc;
  }, []);

export const tryParseJson = (json: string | Error) => {
  let result;

  try {
    result = json.toString ? JSON.parse(json.toString()) : json;
  } catch (error) {
    result = json;
  }

  return result;
};

export const lengthToDuration = (length: string) => {
  const segments = length.split(":");

  switch (segments.length) {
    default:
    case 2:
      return `PT${segments[0]}M${segments[1]}S`;

    case 3:
      return `PT${segments[0]}H${segments[1]}M${segments[2]}S`;
  }
};

export const absUrl = (path: string) =>
  process.env.NODE_ENV !== "production"
    ? `http://localhost:${process.env.PORT || window.location.port}${path}`
    : `http://${
        typeof window === "undefined"
          ? process.env.DOMAIN
          : window.location.host
      }${path}`;

export const extractDomain = (url: string) => url.split("/")[2];
