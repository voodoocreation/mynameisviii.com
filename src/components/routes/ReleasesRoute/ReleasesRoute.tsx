import cn from "classnames";
import Head from "next/head";
import * as React from "react";
import { FormattedMessage, InjectedIntlProps } from "react-intl";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";

import { absoluteUrl, s3ThemeUrl } from "../../../helpers/dataTransformers";
import injectIntlIntoPage from "../../../helpers/injectIntlIntoPage";
import { IRelease } from "../../../models/root.models";
import { TStoreState } from "../../../reducers/root.reducers";
import { IPageContext } from "../../connected/App/App";
import OfflineNotice from "../../connected/OfflineNotice/OfflineNotice";
import ButtonBar from "../../presentation/ButtonBar/ButtonBar";
import LoadButton from "../../presentation/LoadButton/LoadButton";
import NoResults from "../../presentation/NoResults/NoResults";
import PageHeader from "../../presentation/PageHeader/PageHeader";
import ReleaseListing from "../../presentation/ReleaseListing/ReleaseListing";

import * as actions from "../../../actions/root.actions";
import * as selectors from "../../../selectors/root.selectors";

import "./ReleasesRoute.scss";

interface IProps extends InjectedIntlProps {
  fetchMoreReleases: typeof actions.fetchMoreReleases.started;
  fetchReleases: typeof actions.fetchReleases.started;
  hasAllReleases: boolean;
  hasError: boolean;
  isLoading: boolean;
  releases: Record<string, IRelease[]>;
  releasesCount: number;
}

interface IState {
  loadedListings: Record<string, boolean>;
}

class ReleasesRoute extends React.Component<IProps, IState> {
  public static async getInitialProps(context: IPageContext) {
    const { isServer, store } = context;

    const state = store.getState();

    if (isServer || selectors.getReleasesCount(state) === 0) {
      store.dispatch(actions.fetchReleases.started({}));
    } else if (!selectors.getHasAllReleases(state)) {
      store.dispatch(actions.fetchMoreReleases.started({}));
    }
  }

  public readonly state: IState = {
    loadedListings: {}
  };

  public render() {
    const { hasAllReleases, hasError, releases, releasesCount } = this.props;
    const { formatMessage } = this.props.intl;

    const hasLoadedAllListings =
      releasesCount === Object.keys(this.state.loadedListings).length;

    const isLoading = this.props.isLoading || !hasLoadedAllListings;

    const pageTitle = formatMessage({ id: "RELEASES_TITLE" });
    const pageDescription = formatMessage({ id: "RELEASES_DESCRIPTION" });

    return (
      <article className={cn("ReleasesRoute", { hasLoadedAllListings })}>
        <Head>
          <title>
            {pageTitle}
            {" Â· "}
            {formatMessage({ id: "BRAND_NAME" })}
          </title>

          <meta content={pageDescription} name="description" />
          <meta property="og:title" content={pageTitle} />
          <meta property="og:description" content={pageDescription} />
          <meta property="og:type" content="website" />
          <meta property="og:image" content={s3ThemeUrl("/og/releases.jpg")} />
          <meta property="og:url" content={absoluteUrl("/releases")} />
        </Head>

        <PageHeader>
          <FormattedMessage id="RELEASES_TITLE" />
        </PageHeader>

        {hasError ? <OfflineNotice /> : null}

        {releasesCount > 0
          ? Object.keys(releases).map(this.renderListingsForType)
          : null}

        {hasAllReleases && releasesCount === 0 ? (
          <NoResults>
            <p>
              <FormattedMessage id="NO_RELEASES" />
            </p>
          </NoResults>
        ) : null}

        <ButtonBar>
          {!hasAllReleases ? (
            <LoadButton
              hasError={hasError}
              isLoading={isLoading}
              onLoad={this.onLoadMore}
            />
          ) : null}
        </ButtonBar>
      </article>
    );
  }

  private renderListingsForType = (type: string) =>
    this.props.releases[type] ? (
      <section
        className={cn(
          "ReleasesRoute--listings",
          `ReleasesRoute--listings-${type}`
        )}
        key={type}
      >
        <h2>
          <FormattedMessage
            id={type.toUpperCase()}
            values={{ count: this.props.releases[type].length }}
          />
        </h2>

        <div className="ReleasesRoute--listings--items">
          {this.props.releases[type].map((release: IRelease) => (
            <ReleaseListing
              {...release}
              key={release.slug}
              onLoad={this.onListingLoad(release)}
            />
          ))}
        </div>
      </section>
    ) : null;

  private onListingLoad = (release: IRelease) => () => {
    this.setState({
      loadedListings: {
        ...this.state.loadedListings,
        [release.slug]: true
      }
    });
  };

  private onLoadMore = () => {
    this.props.fetchMoreReleases({});
  };
}

const mapState = (state: TStoreState) => ({
  hasAllReleases: selectors.getHasAllReleases(state),
  hasError: selectors.hasReleasesError(state),
  isLoading: selectors.getReleasesIsLoading(state),
  releases: selectors.getSortedReleasesByType(state),
  releasesCount: selectors.getReleasesCount(state)
});

const mapActions = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      fetchMoreReleases: actions.fetchMoreReleases.started,
      fetchReleases: actions.fetchReleases.started
    },
    dispatch
  );

export default injectIntlIntoPage(
  connect(
    mapState,
    mapActions
  )(ReleasesRoute)
);
