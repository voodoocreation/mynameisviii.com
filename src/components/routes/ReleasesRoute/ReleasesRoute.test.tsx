import { mount, render } from "enzyme";
import { camelizeKeys } from "humps";
import merge from "lodash.merge";
import * as React from "react";
import { Provider } from "react-redux";

import * as selectors from "../../../selectors/root.selectors";
import createStore from "../../../store/root.store";
import ReleasesRoute from "./ReleasesRoute";

import releases from "../../../../server/mocks/releases.json";
import * as dom from "../../../helpers/dom";
import { arrayToAssoc } from "../../../transformers/transformData";

const mockData: any = camelizeKeys(releases);

Object.defineProperty(dom, "isAlmostInViewport", {
  value: jest.fn(() => false)
});

const setup = (fn: any, fromTestStore = {}, fromTestApi?: {}) => {
  const store = createStore(
    merge(
      {
        releases: {
          items: {}
        }
      },
      fromTestStore
    ),
    {},
    fromTestApi
  );

  return {
    actual: fn(
      <Provider store={store}>
        <ReleasesRoute />
      </Provider>
    ),
    store
  };
};

describe("[routes] <ReleasesRoute />", () => {
  it("renders correctly", () => {
    const { actual } = setup(render, {
      releases: {
        hasAllItems: true,
        items: arrayToAssoc(mockData.items, "slug")
      }
    });

    expect(actual).toMatchSnapshot();
  });

  it("renders no results correctly", () => {
    const { actual } = setup(render, {
      releases: {
        hasAllItems: true
      }
    });

    expect(actual.find(".NoResults")).toHaveLength(1);
  });

  it("renders with an error correctly", () => {
    const { actual } = setup(render, {
      releases: {
        error: {
          message: "Server error",
          status: 500
        }
      }
    });

    expect(actual.find(".LoadButton").text()).toBe("Try again");
    expect(actual).toMatchSnapshot();
  });

  it("renders with 'load more' button when hasAllItems=false", () => {
    const { actual } = setup(render, {
      releases: {
        hasAllItems: false
      }
    });

    expect(actual.find(".LoadButton")).toHaveLength(1);
  });

  it("renders without 'load more' button when hasAllItems=true", () => {
    const { actual } = setup(render, {
      releases: {
        hasAllItems: true
      }
    });

    expect(actual.find(".LoadButton")).toHaveLength(0);
  });

  it("loads more listings when 'load more' button is activated", () => {
    const { actual } = setup(
      mount,
      {},
      {
        fetchReleases: () => ({
          data: {
            items: [mockData.items[0]]
          },
          ok: true
        })
      }
    );

    expect(actual.find("ReleaseListing")).toHaveLength(0);
    actual.find("LoadButton").simulate("click");
    expect(actual.find("ReleaseListing")).toHaveLength(1);
  });

  it("updates `loadedListings` state after a listing has loaded", () => {
    const { actual } = setup(mount, {
      releases: {
        items: { "test-1": mockData.items[0] }
      }
    });

    actual.find("ReleaseListing img").simulate("load");

    expect(actual.find(".hasLoadedAllListings")).toHaveLength(1);
  });

  describe("getInitialProps()", () => {
    const items = [mockData.items[0]];

    it("fetches items when rendering on the server", async () => {
      const { store } = setup(
        render,
        {},
        {
          fetchReleases: () => ({
            data: { items },
            ok: true
          })
        }
      );

      await ReleasesRoute.getInitialProps({
        ctx: {
          isServer: true,
          store
        }
      });

      expect(selectors.getReleasesAsArray(store.getState())).toEqual(items);
    });

    it("fetches items when the store is empty", async () => {
      const { store } = setup(
        render,
        {},
        {
          fetchReleases: () => ({
            data: { items },
            ok: true
          })
        }
      );

      await ReleasesRoute.getInitialProps({
        ctx: {
          isServer: false,
          store
        }
      });

      expect(selectors.getReleasesAsArray(store.getState())).toEqual(items);
    });

    it("fetches more items when there are more to fetch", async () => {
      const { store } = setup(
        render,
        {
          releases: {
            items: {
              "test-1": items[0]
            }
          }
        },
        {
          fetchReleases: () => ({
            data: {
              items: [{ slug: "test-2" }]
            },
            ok: true
          })
        }
      );

      expect(selectors.getReleasesCount(store.getState())).toBe(1);

      await ReleasesRoute.getInitialProps({
        ctx: {
          isServer: false,
          store
        }
      });

      expect(selectors.getReleasesCount(store.getState())).toBe(2);
    });

    it("doesn't fetch anything when all items are in the store", async () => {
      const { store } = setup(render, {
        releases: {
          hasAllItems: true,
          items: {
            "test-1": items[0]
          }
        }
      });

      await ReleasesRoute.getInitialProps({
        ctx: {
          isServer: false,
          store
        }
      });

      expect(selectors.getReleasesAsArray(store.getState())).toEqual(items);
    });
  });
});
