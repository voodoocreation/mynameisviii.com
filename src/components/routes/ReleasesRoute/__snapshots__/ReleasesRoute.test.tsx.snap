// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[routes] <ReleasesRoute /> renders correctly 1`] = `
<article
  class="ReleasesRoute"
>
  <header
    class="PageHeader"
  >
    <h1>
      <span>
        Releases
      </span>
    </h1>
  </header>
  <section
    class="ReleaseListings ReleaseListings-single"
  >
    <h2>
      <span>
        Singles
      </span>
    </h2>
    <div
      class="ReleaseListings-items"
    >
      <article
        class="ReleaseListing"
      >
        <a
          href="/releases/lovedecay-lovesong"
        >
          <div
            class="ReleaseListing-details"
          >
            <header
              class="ReleaseListing-header"
            >
              <h3>
                <span>
                  lovedecay / lovesong
                </span>
              </h3>
            </header>
            <section
              class="ReleaseListing-meta"
            >
              <div
                class="Meta ReleaseListing-releasedOn"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m31.6 33.4v-18.4h-23.2v18.4h23.2z m0-26.8c1.8 0 3.4 1.6 3.4 3.4v23.4c0 1.8-1.6 3.2-3.4 3.2h-23.2c-1.9 0-3.4-1.4-3.4-3.2v-23.4c0-1.8 1.5-3.4 3.4-3.4h1.6v-3.2h3.4v3.2h13.2v-3.2h3.4v3.2h1.6z m-3.2 11.8v3.2h-3.4v-3.2h3.4z m-6.8 0v3.2h-3.2v-3.2h3.2z m-6.6 0v3.2h-3.4v-3.2h3.4z"
                    />
                  </g>
                </svg>
                 
                <span
                  class="Meta-label"
                >
                  <span>
                    Released
                  </span>
                  :
                </span>
                 
                <span
                  class="Meta-value"
                >
                  <time
                    class="DateTime"
                    datetime="2017-03-23T11:00:00.000Z"
                    title="March 24, 2017"
                  >
                    March 24, 2017
                  </time>
                </span>
              </div>
              <div
                class="Meta ReleaseListing-length"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m20.9 11.6v8.8l7.5 4.4-1.3 2.2-8.7-5.4v-10h2.5z m-0.9 21.8q5.5 0 9.4-4t4-9.4-4-9.4-9.4-4-9.4 4-4 9.4 4 9.4 9.4 4z m0-30q6.9 0 11.8 4.8t4.8 11.8-4.8 11.8-11.8 4.8-11.8-4.8-4.8-11.8 4.8-11.8 11.8-4.8z"
                    />
                  </g>
                </svg>
                 
                <span
                  class="Meta-label"
                >
                  <span>
                    Length
                  </span>
                  :
                </span>
                 
                <span
                  class="Meta-value"
                >
                  7:31
                </span>
              </div>
              <div
                class="Meta ReleaseListing-genre"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m20 5h10v6.6h-6.6v16.8c0 3.6-3.1 6.6-6.8 6.6s-6.6-3-6.6-6.6 3-6.8 6.6-6.8c1.3 0 2.4 0.4 3.4 1v-17.6z"
                    />
                  </g>
                </svg>
                 
                <span
                  class="Meta-label"
                >
                  <span>
                    Genre
                  </span>
                  :
                </span>
                 
                <span
                  class="Meta-value"
                >
                  Goth/Electronic
                </span>
              </div>
              <div
                class="Meta ReleaseListing-tracks"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m11.6 21.6v-3.2h23.4v3.2h-23.4z m0 10v-3.2h23.4v3.2h-23.4z m0-23.2h23.4v3.2h-23.4v-3.2z m-8.2 10v-1.8h5v1.6l-3.1 3.4h3.1v1.8h-5v-1.6l2.9-3.4h-2.9z m1.6-5v-5h-1.6v-1.8h3.2v6.8h-1.6z m-1.6 15v-1.8h5v6.8h-5v-1.8h3.2v-0.7h-1.6v-1.8h1.6v-0.7h-3.2z"
                    />
                  </g>
                </svg>
                  
                <span
                  class="Meta-value"
                >
                  2 
                  <span>
                    tracks
                  </span>
                </span>
              </div>
            </section>
          </div>
          <img
            alt="lovedecay / lovesong"
            src="https://s3.amazonaws.com/mynameisviii-static/releases/lovedecay-lovesong.jpg"
          />
        </a>
      </article>
      <article
        class="ReleaseListing"
      >
        <a
          href="/releases/heavens-to-elizabeth"
        >
          <div
            class="ReleaseListing-details"
          >
            <header
              class="ReleaseListing-header"
            >
              <h3>
                <span>
                  Heavens to Elizabeth
                </span>
              </h3>
            </header>
            <section
              class="ReleaseListing-meta"
            >
              <div
                class="Meta ReleaseListing-releasedOn"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m31.6 33.4v-18.4h-23.2v18.4h23.2z m0-26.8c1.8 0 3.4 1.6 3.4 3.4v23.4c0 1.8-1.6 3.2-3.4 3.2h-23.2c-1.9 0-3.4-1.4-3.4-3.2v-23.4c0-1.8 1.5-3.4 3.4-3.4h1.6v-3.2h3.4v3.2h13.2v-3.2h3.4v3.2h1.6z m-3.2 11.8v3.2h-3.4v-3.2h3.4z m-6.8 0v3.2h-3.2v-3.2h3.2z m-6.6 0v3.2h-3.4v-3.2h3.4z"
                    />
                  </g>
                </svg>
                 
                <span
                  class="Meta-label"
                >
                  <span>
                    Released
                  </span>
                  :
                </span>
                 
                <span
                  class="Meta-value"
                >
                  <time
                    class="DateTime"
                    datetime="2017-05-15T11:00:00.000Z"
                    title="May 15, 2017"
                  >
                    May 15, 2017
                  </time>
                </span>
              </div>
              <div
                class="Meta ReleaseListing-length"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m20.9 11.6v8.8l7.5 4.4-1.3 2.2-8.7-5.4v-10h2.5z m-0.9 21.8q5.5 0 9.4-4t4-9.4-4-9.4-9.4-4-9.4 4-4 9.4 4 9.4 9.4 4z m0-30q6.9 0 11.8 4.8t4.8 11.8-4.8 11.8-11.8 4.8-11.8-4.8-4.8-11.8 4.8-11.8 11.8-4.8z"
                    />
                  </g>
                </svg>
                 
                <span
                  class="Meta-label"
                >
                  <span>
                    Length
                  </span>
                  :
                </span>
                 
                <span
                  class="Meta-value"
                >
                  11:09
                </span>
              </div>
              <div
                class="Meta ReleaseListing-genre"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m20 5h10v6.6h-6.6v16.8c0 3.6-3.1 6.6-6.8 6.6s-6.6-3-6.6-6.6 3-6.8 6.6-6.8c1.3 0 2.4 0.4 3.4 1v-17.6z"
                    />
                  </g>
                </svg>
                 
                <span
                  class="Meta-label"
                >
                  <span>
                    Genre
                  </span>
                  :
                </span>
                 
                <span
                  class="Meta-value"
                >
                  Goth/Rock
                </span>
              </div>
              <div
                class="Meta ReleaseListing-tracks"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m11.6 21.6v-3.2h23.4v3.2h-23.4z m0 10v-3.2h23.4v3.2h-23.4z m0-23.2h23.4v3.2h-23.4v-3.2z m-8.2 10v-1.8h5v1.6l-3.1 3.4h3.1v1.8h-5v-1.6l2.9-3.4h-2.9z m1.6-5v-5h-1.6v-1.8h3.2v6.8h-1.6z m-1.6 15v-1.8h5v6.8h-5v-1.8h3.2v-0.7h-1.6v-1.8h1.6v-0.7h-3.2z"
                    />
                  </g>
                </svg>
                  
                <span
                  class="Meta-value"
                >
                  3 
                  <span>
                    tracks
                  </span>
                </span>
              </div>
            </section>
          </div>
          <img
            alt="Heavens to Elizabeth"
            src="https://s3.amazonaws.com/mynameisviii-static/releases/heavens-to-elizabeth.jpg"
          />
        </a>
      </article>
      <article
        class="ReleaseListing"
      >
        <a
          href="/releases/witchlust-rev-22-20"
        >
          <div
            class="ReleaseListing-details"
          >
            <header
              class="ReleaseListing-header"
            >
              <h3>
                <span>
                  witchlust / rev 22:20
                </span>
              </h3>
            </header>
            <section
              class="ReleaseListing-meta"
            >
              <div
                class="Meta ReleaseListing-releasedOn"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m31.6 33.4v-18.4h-23.2v18.4h23.2z m0-26.8c1.8 0 3.4 1.6 3.4 3.4v23.4c0 1.8-1.6 3.2-3.4 3.2h-23.2c-1.9 0-3.4-1.4-3.4-3.2v-23.4c0-1.8 1.5-3.4 3.4-3.4h1.6v-3.2h3.4v3.2h13.2v-3.2h3.4v3.2h1.6z m-3.2 11.8v3.2h-3.4v-3.2h3.4z m-6.8 0v3.2h-3.2v-3.2h3.2z m-6.6 0v3.2h-3.4v-3.2h3.4z"
                    />
                  </g>
                </svg>
                 
                <span
                  class="Meta-label"
                >
                  <span>
                    Released
                  </span>
                  :
                </span>
                 
                <span
                  class="Meta-value"
                >
                  <time
                    class="DateTime"
                    datetime="2017-06-01T11:00:00.000Z"
                    title="June 1, 2017"
                  >
                    June 1, 2017
                  </time>
                </span>
              </div>
              <div
                class="Meta ReleaseListing-length"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m20.9 11.6v8.8l7.5 4.4-1.3 2.2-8.7-5.4v-10h2.5z m-0.9 21.8q5.5 0 9.4-4t4-9.4-4-9.4-9.4-4-9.4 4-4 9.4 4 9.4 9.4 4z m0-30q6.9 0 11.8 4.8t4.8 11.8-4.8 11.8-11.8 4.8-11.8-4.8-4.8-11.8 4.8-11.8 11.8-4.8z"
                    />
                  </g>
                </svg>
                 
                <span
                  class="Meta-label"
                >
                  <span>
                    Length
                  </span>
                  :
                </span>
                 
                <span
                  class="Meta-value"
                >
                  8:54
                </span>
              </div>
              <div
                class="Meta ReleaseListing-genre"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m20 5h10v6.6h-6.6v16.8c0 3.6-3.1 6.6-6.8 6.6s-6.6-3-6.6-6.6 3-6.8 6.6-6.8c1.3 0 2.4 0.4 3.4 1v-17.6z"
                    />
                  </g>
                </svg>
                 
                <span
                  class="Meta-label"
                >
                  <span>
                    Genre
                  </span>
                  :
                </span>
                 
                <span
                  class="Meta-value"
                >
                  Goth/Electronic
                </span>
              </div>
              <div
                class="Meta ReleaseListing-tracks"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m11.6 21.6v-3.2h23.4v3.2h-23.4z m0 10v-3.2h23.4v3.2h-23.4z m0-23.2h23.4v3.2h-23.4v-3.2z m-8.2 10v-1.8h5v1.6l-3.1 3.4h3.1v1.8h-5v-1.6l2.9-3.4h-2.9z m1.6-5v-5h-1.6v-1.8h3.2v6.8h-1.6z m-1.6 15v-1.8h5v6.8h-5v-1.8h3.2v-0.7h-1.6v-1.8h1.6v-0.7h-3.2z"
                    />
                  </g>
                </svg>
                  
                <span
                  class="Meta-value"
                >
                  2 
                  <span>
                    tracks
                  </span>
                </span>
              </div>
            </section>
          </div>
          <img
            alt="witchlust / rev 22:20"
            src="https://s3.amazonaws.com/mynameisviii-static/releases/witchlust-rev-22-20.jpg"
          />
        </a>
      </article>
      <article
        class="ReleaseListing"
      >
        <a
          href="/releases/threadfear-ask-me-to-stay"
        >
          <div
            class="ReleaseListing-details"
          >
            <header
              class="ReleaseListing-header"
            >
              <h3>
                <span>
                  threadfear / ask me to stay
                </span>
              </h3>
            </header>
            <section
              class="ReleaseListing-meta"
            >
              <div
                class="Meta ReleaseListing-releasedOn"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m31.6 33.4v-18.4h-23.2v18.4h23.2z m0-26.8c1.8 0 3.4 1.6 3.4 3.4v23.4c0 1.8-1.6 3.2-3.4 3.2h-23.2c-1.9 0-3.4-1.4-3.4-3.2v-23.4c0-1.8 1.5-3.4 3.4-3.4h1.6v-3.2h3.4v3.2h13.2v-3.2h3.4v3.2h1.6z m-3.2 11.8v3.2h-3.4v-3.2h3.4z m-6.8 0v3.2h-3.2v-3.2h3.2z m-6.6 0v3.2h-3.4v-3.2h3.4z"
                    />
                  </g>
                </svg>
                 
                <span
                  class="Meta-label"
                >
                  <span>
                    Released
                  </span>
                  :
                </span>
                 
                <span
                  class="Meta-value"
                >
                  <time
                    class="DateTime"
                    datetime="2017-11-02T11:00:00.000Z"
                    title="November 3, 2017"
                  >
                    November 3, 2017
                  </time>
                </span>
              </div>
              <div
                class="Meta ReleaseListing-length"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m20.9 11.6v8.8l7.5 4.4-1.3 2.2-8.7-5.4v-10h2.5z m-0.9 21.8q5.5 0 9.4-4t4-9.4-4-9.4-9.4-4-9.4 4-4 9.4 4 9.4 9.4 4z m0-30q6.9 0 11.8 4.8t4.8 11.8-4.8 11.8-11.8 4.8-11.8-4.8-4.8-11.8 4.8-11.8 11.8-4.8z"
                    />
                  </g>
                </svg>
                 
                <span
                  class="Meta-label"
                >
                  <span>
                    Length
                  </span>
                  :
                </span>
                 
                <span
                  class="Meta-value"
                >
                  9:41
                </span>
              </div>
              <div
                class="Meta ReleaseListing-genre"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m20 5h10v6.6h-6.6v16.8c0 3.6-3.1 6.6-6.8 6.6s-6.6-3-6.6-6.6 3-6.8 6.6-6.8c1.3 0 2.4 0.4 3.4 1v-17.6z"
                    />
                  </g>
                </svg>
                 
                <span
                  class="Meta-label"
                >
                  <span>
                    Genre
                  </span>
                  :
                </span>
                 
                <span
                  class="Meta-value"
                >
                  Goth/Rock
                </span>
              </div>
              <div
                class="Meta ReleaseListing-tracks"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="vertical-align:middle"
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m11.6 21.6v-3.2h23.4v3.2h-23.4z m0 10v-3.2h23.4v3.2h-23.4z m0-23.2h23.4v3.2h-23.4v-3.2z m-8.2 10v-1.8h5v1.6l-3.1 3.4h3.1v1.8h-5v-1.6l2.9-3.4h-2.9z m1.6-5v-5h-1.6v-1.8h3.2v6.8h-1.6z m-1.6 15v-1.8h5v6.8h-5v-1.8h3.2v-0.7h-1.6v-1.8h1.6v-0.7h-3.2z"
                    />
                  </g>
                </svg>
                  
                <span
                  class="Meta-value"
                >
                  2 
                  <span>
                    tracks
                  </span>
                </span>
              </div>
            </section>
          </div>
          <img
            alt="threadfear / ask me to stay"
            src="https://s3.amazonaws.com/mynameisviii-static/releases/threadfear-ask-me-to-stay.jpg"
          />
        </a>
      </article>
    </div>
  </section>
</article>
`;
